File: CODE/T4006/T4006.cpp
#define nametask "T4006"
#include <bits/stdc++.h>
#define fi first
#define se second
#define mp make_pair
#define pb push_back
#define SZ(a) (int) a.size()
#define all(a) a.begin(), a.end()
#define FOR(i, a, b) for (int i = a; i <= b; i++)
#define FOD(i, b, a) for (int i = b; i >= a; i--)

using namespace std;
typedef long long ll;
typedef pair <int, int> pi;
const int MOD = 1e9 + 7;

string n;
int f[1005][3];

void add(int &x, int y){
    x += y;
    if (x >= MOD) x -= MOD;
}

int calc(int id, bool tight, int rem, int sz){
    if (id >= sz)
        return ((rem == 0) && (tight == 0));
    if ((!tight) && (f[id][rem] != -1)) return f[id][rem];
    int cnt = 0;
    int maxc = (tight) ? (n[id] - '0') : 9;
    FOR(i, 0, maxc){
        int new_tight = (i == (n[id] - '0')) ? tight : 0;
        add(cnt, calc(id + 1, new_tight, (rem + i * i) % 3, sz));
    }
    if (!tight)
        f[id][rem] = cnt;
    return cnt;
}

void solve(){
    memset(f, -1, sizeof f);
    cin >> n;
    int sz = SZ(n);
    cout << calc(0, 1, 0, sz);
}

signed main()
{
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    if (fopen(nametask".inp", "r")){
        freopen(nametask".inp", "r", stdin);
        freopen(nametask".out", "w", stdout);
    }
    solve();
    return 0;
}

File: CODE/T4006/T4006.py
import sys

sys.stdin = open("T4006.inp", "r")
sys.stdout = open("T4006.out", "w")

MOD = int(1e9 + 7)

def add(x, y):
    return (x + y) % MOD

def digit_dp(idx, tight, rem, digits, dp):
    if idx == len(digits):
        return 1 if (rem == 0 and not tight) else 0
    if not tight and dp[idx][rem] != -1:
        return dp[idx][rem]
    
    max_digit = digits[idx] if tight else 9
    cnt = 0
    
    for i in range(max_digit + 1):
        new_tight = tight and (i == digits[idx])
        cnt = add(cnt, digit_dp(idx + 1, new_tight, (rem + i * i) % 3, digits, dp))
    
    if not tight:
        dp[idx][rem] = cnt
    return cnt

def count_divisible_numbers(n):
    digits = list(map(int, n))
    dp = [[-1] * 3 for _ in range(len(digits))]
    return digit_dp(0, True, 0, digits, dp)

n = input()
print(count_divisible_numbers(n))
